Remove addgroup from user

jvillafanez opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.com/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next version of ownCloud.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" if the PR still has open tasks
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
This PR includes 3 changes:
* Remove the "add group" button from the dropdowns. New groups can be added from the "add group" button in the top left side of the page. The dropdowns will only be used to select or deselect groups.
* Fix issue with the group count in the left sidebar when a new group was created (group count still shows only if ldap is disabled)
* Use group's displayname in more places in the web UI

The 3rd point should be mostly transparent. The places where the displayname wasn't used are mostly pointing to local groups, whose id and displayname matches. There are still a couple of messages using the group id, but those will require more changes.

## Related Issue
https://github.com/owncloud/enterprise/issues/5684

## Motivation and Context
The "add group" button was removed because the behavior could cause confusion, specially when creating a new user: the groups weren't created until the user was created, but they appear as such in the web UI.

The changes in the displayname shouldn't be noticed, but the web UI should be more prepared now

## How Has This Been Tested?


## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

jvillafanez committed:
Remove add group option from the dropdown menus

jvillafanez committed:
Fix user count after adding a new user to a group

jvillafanez committed:
Use the group's displayname in more places

jvillafanez committed:
Homogenize group data

jvillafanez committed:
Fix tests and adjust code

jvillafanez committed:
Increase coverage

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiWebdavLocksUnlock-maria10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38379/84
<!-- id: pr-40770 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiFederationToRoot1-latest-maria10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38379/97
<!-- id: pr-40770 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiWebdavMove2-maria10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38379/86
<!-- id: pr-40770 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiWebdavUpload1-maria10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38379/91
<!-- id: pr-40770 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiSharingNotificationsToRoot-maria10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38379/94
<!-- id: pr-40770 -->

jvillafanez committed:
Add changelog entry

jvillafanez commented:
In order to check the changes for the displayname, you can use the following patch (*)
```
diff --git a/lib/private/Group/Database.php b/lib/private/Group/Database.php
index 4b8eb79de3..daf946583e 100644
--- a/lib/private/Group/Database.php
+++ b/lib/private/Group/Database.php
@@ -349,4 +349,11 @@ class Database extends \OC\Group\Backend {
                }
                return $count;
        }
+
+       public function getGroupDetails($gid) {
+               return [
+                       'gid' => $gid,
+                       'displayName' => "{$gid}__{$gid}",
+               ];
+       }
 }
```
(*) This is intended to be used just for testing. DO NOT use it on production. Right now, it isn't possible to use a displayname for local groups, and there is no short term plan for this feature.

jvillafanez commented:
There are no tests for js files, and the ajax part can't be unit tested, so it isn't possible to rise the code coverage.

pako81 commented:
It would nice to also have this in 10.13. @phil-davis @jnweiger what to do here?

phil-davis commented:
> It would be nice to also have this in 10.13. @phil-davis @jnweiger what to do here?

IMO this is "a good thing". I don't think that it is really necessary to be able to add a group on-the-fly while adding a user. Groups can be easily created "separately", which is how I imagine it is done 99% of the time anyway. Removing the "add group" stuff from inside the "add user" UI means we can avoid all the edge cases for what happens when the admin has got to various points in the UI workflow, then aborted, or "something happened" when trying to save the user (and new group along with it)...

We just need a developer review. Who do we mention for reviews of PRs done by @jvillafanez ?

pako81 commented:
> We just need a developer review. Who do we mention for reviews of PRs done by @jvillafanez ?

Had a look at the code myself and it looks ok. Still it would be good if someone else could review this. @mrow4a maybe?

phil-davis commented:
Note: existing UI automated acceptance tests cover adding a group from the "normal" left-hand side of the admin "users page". They also cover adding a user, including selecting existing groups that the user will become a member of. VI passes, so the behavior of all that "ordinary workflow: still works.

We never had automated acceptance tests that created a group on-the-fly while creating a user. So that is why there are no acceptance tests to remove when removing that functionality.

cdamken commented:
> Had a look at the code myself and it looks ok. Still it would be good if someone else could review this. @mrow4a or @steelcuts maybe?

I asked both if they can provide a review.

mrow4a commented on settings/js/users/groups.js:
```diff
@@ -24,11 +24,11 @@ var GroupList;
 				'<li class="isgroup" data-gid="' + gid + '" data-usercount="0">' +
 				'	<a href="#" class="dorename">' +
-				'		<span class="groupname">' + gid + '</span>' +
+				'		<span class="groupname">' + name + '</span>' +
```
good idea

mrow4a commented on settings/Controller/GroupsController.php:
```diff
@@ -175,23 +176,42 @@ public function getAssignableAndRemovableGroups() {
+					$groupObject = $this->groupManager->get($group);
+					if ($backend->implementsActions($backend::ADD_TO_GROUP)) {
+						$assignableGroups[$groupObject->getGID()] = [
+							'id' => $groupObject->getGID(),
```
id -> gid

mrow4a commented on settings/js/users/users.js:
```diff
@@ -494,17 +494,21 @@ var UserList = {
 
 							if (response.data.action === 'add') {
-								GroupList.incGroupCount(groupName);
+								GroupList.incGroupCount(groupId);
```
groupGID

mrow4a commented on settings/js/users/users.js:
```diff
@@ -706,31 +678,31 @@ var UserList = {
 		}
 
-		function createItem(group) {
+		function createItem(id, group) {
```
createGroupItem(gid, name)

mrow4a reviewed (commented):
minor clarifications, plus what do we do with Sonar?

jvillafanez committed:
Replace id to gid

jvillafanez commented:
I've replaced the "id" to "gid". I've also added some changes in the last commit in "settings/js/users/users.js" around lines 737 and 759 in order to retrieve the group info, which seemed to give some problems in some scenarios (the previous behavior might be okish, but there were some errors in the console).

pako81 commented:
Acceptance tests are failing with `Page\UsersPage::waitTillPageIsLoaded timeout waiting for user list to load on users page (Exception)` --> https://github.com/owncloud/core/blob/master/tests/acceptance/features/lib/UsersPage.php#L982

phil-davis committed:
Do not use null coalescing operator in JS

phil-davis commented:
The UI acceptance tests run with an old version of Chrome browser. That is because of issues with the Behat-Mink-selenium test framework, it is not able to control newer versions of Chrome or Firefox that have a newer interface for automated testing.
We have not pushed to try and fix this upstream, because the Behat-Mink-Selenium stack is only being used for automated UI testing of the classic web UI, and that does not have much change happening.

But it means that new JS features cannot be used. In this case, the "null coalescing operator" `??` is a "new" JavaScript thing. It is not valid in the old version of Chrome used in the automated tests.

I changed the syntax back from `x ?? y` to `x ? x : y` syntax. See the recent commit that I pushed. The tests should pass.

phil-davis commented:
> There are no tests for js files, and the ajax part can't be unit tested, so it isn't possible to rise the code coverage.

Note: we know why SonarCloud is complaining about the code coverage.

IMO this can be merged - drone CI passed.
